### PR TITLE
add feature to support SASL for Kafka and MS EventHub

### PR DESCRIPTION
**Pull request by lmtyler**

### CODE REVIEW

File: docs/config.md [hunk 1]
```diff
@@ -53,6 +54,19 @@ NATS, streaming, Kafka and HTTP configurations all take an optional TLS setting.
 * `cert` - file path to a server certificate, used for HTTPS monitoring and optionally for client side certificates with NATS
 * `key` - key for the certificate store specified in cert
 
+## SASL <a name="sasl"></a>
+
+Kafka allow for SASL_PLAIN connection with user and password.  This also supports a flag to connect to Azure EventHub
```
Comment (derekcollison) [inline docs/config.md:59]: allows

File: resources/certs/ca-cert [hunk 2]
```diff
@@ -1,20 +1,20 @@
 -----BEGIN CERTIFICATE-----
-MIIDRjCCAi4CCQCZ2pXVwpC9gzANBgkqhkiG9w0BAQsFADBlMQswCQYDVQQGEwJV
```
Comment (derekcollison) [inline resources/certs/ca-cert:2]: Why did we change the CA cert?

File: resources/jaas/kafka_server_jaas.conf [hunk 3]
```diff
@@ -0,0 +1,8 @@
+KafkaServer {
+  org.apache.kafka.common.security.plain.PlainLoginModule required
+  username="admin"
+  password="admin-secret"
+  user_admin="admin-secret"
```
Comment (derekcollison) [inline resources/jaas/kafka_server_jaas.conf:5]: These necessary? Are they testing another way to configure user/pass?

File: resources/test_servers.yml [hunk 4]
```diff
@@ -1,27 +1,34 @@
 version: '2'
 services:
   zookeeper:
-    image: wurstmeister/zookeeper
+    image: hub.docker.prod.walmart.com/wurstmeister/zookeeper
```
Comment (derekcollison) [inline resources/test_servers.yml:4]: You want yours here or public one?

File: server/core/connector.go [hunk 5]
```diff
@@ -132,7 +135,21 @@ func (conn *BridgeConnector) initKafka() {
 		Timeout:   time.Duration(conn.bridge.config.ConnectTimeout) * time.Millisecond,
 		DualStack: true,
 	}
-	tlsC, err := conn.config.TLS.MakeTLSConfig()
+	var err error
+	var tlsC *tls.Config
+	if conn.config.SASL.User == "" && conn.config.SASL.Password == "" {
```
Comment (derekcollison) [inline server/core/connector.go:140]: Can empty password be legit (no comment on security)?

Comment (derekcollison) [reply server/core/connector.go:140]: Meaning should we just check SASL user?

**Review by derekcollison — COMMENTED** [review]
In general looks good, a few questions/comments.

File: resources/certs/ca-cert (hunk 2 above)
Comment (lmtyler) [reply resources/certs/ca-cert:2]: The TLS tests did not work without updating the certs.  Assumed they are expired, but can put them back to be regened locally for each user.

File: resources/test_servers.yml (hunk 4 above)
Comment (lmtyler) [reply resources/test_servers.yml:4]: Sorry, I missed this one and will put it back to public.

File: server/core/connector.go (hunk 5 above)
Comment (lmtyler) [reply server/core/connector.go:140]: Fair question, don't think a blank password is legit.  Will change the check to only be for SASL userid.

File: resources/certs/ca-cert (hunk 2 above)
Comment (lmtyler) [reply resources/certs/ca-cert:2]: reset back to upstream/master for net no change

File: server/core/connector.go (hunk 5 above)
Comment (lmtyler) [reply server/core/connector.go]: changed, commited, and squashed

File: resources/test_servers.yml (hunk 4 above)
Comment (lmtyler) [reply resources/test_servers.yml]: changed, commited, and squashed

File: resources/jaas/kafka_server_jaas.conf (hunk 3 above)
Comment (lmtyler) [reply resources/jaas/kafka_server_jaas.conf:5]: I looked around and this seemed to be the consistent way of configuring SASL for Kafka

File: resources/certs/ca-cert (hunk 2 above)
Comment (lmtyler) [reply resources/certs/ca-cert]: changed, commited, and squashed

File: resources/jaas/kafka_server_jaas.conf (hunk 3 above)
Comment (lmtyler) [reply resources/jaas/kafka_server_jaas.conf:5]: I looked around and this seems to be the recommended way to configure SASL.

Comment (lmtyler) [reply resources/jaas/kafka_server_jaas.conf:5]: I looked around and this seems to be the recommended way to configure SASL for Kafka

File: resources/certs/ca-cert (hunk 2 above)
Comment (lmtyler) [reply resources/certs/ca-cert]: TLS test were not working so I rebuild them locally.
Reset back to master
commited and squashed
net no change

File: resources/test_servers.yml (hunk 4 above)
Comment (lmtyler) [reply resources/test_servers.yml]: Missed this.
reset back to public docker hub
commited and squashed

File: server/core/connector.go (hunk 5 above)
Comment (lmtyler) [reply server/core/connector.go]: Fair call out.  Don't think it is legit.
Removed check for blank password.
commited and squashed

File: resources/test_servers.yml [hunk 6]
```diff
@@ -9,19 +9,26 @@ services:
...
+      - "9094:9094"
     environment:
       KAFKA_BROKER_ID: 1
       KAFKA_AUTO_CREATE_TOPICS_ENABLE: 'true'
       KAFKA_ADVERTISED_HOST_NAME: localhost
       KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
-      KAFKA_LISTENERS: PLAINTEXT://:9092,SSL://:9093
-      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://localhost:9092,SSL://localhost:9093
+      KAFKA_LISTENERS: PLAINTEXT://:9092,SSL://:9093,SASL_PLAINTEXT://:9094
+      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://localhost:9092,SSL://localhost:9093,SASL_PLAINTEXT://localhost:9094
```
Comment (derekcollison) [inline resources/test_servers.yml:19]: Do we need complete new port for SASL?

File: server/core/server_test.go [hunk 7]
```diff
@@ -21,22 +21,26 @@ import (
...
-	"github.com/nats-io/nats-kafka/server/conf"
 	gnatsserver "github.com/nats-io/nats-server/v2/server"
 	gnatsd "github.com/nats-io/nats-server/v2/test"
 	nss "github.com/nats-io/nats-streaming-server/server"
 	"github.com/nats-io/nats.go"
 	"github.com/nats-io/nuid"
 	"github.com/nats-io/stan.go"
 	"github.com/segmentio/kafka-go"
+	"github.com/segmentio/kafka-go/sasl/plain"
+
```
Comment (derekcollison) [inline server/core/server_test.go:32]: no space

File: server/core/stan2kafka.go [hunk 8]
```diff
@@ -19,8 +19,9 @@ package core
 import (
 	"fmt"
 
+	"github.com/nats-io/stan.go"
+
```
Comment (derekcollison) [inline server/core/stan2kafka.go:23]: no space

**Review by derekcollison — COMMENTED** [review]
We are close, a few minor comments.

File: resources/test_servers.yml (hunk 6 above)
Comment (lmtyler) [reply resources/test_servers.yml:19]: The way I understood the research I did was that the configuration has to be tied to a port in order for Kafka to know how to authenticate.

File: server/core/server_test.go (hunk 7 above)
Comment (lmtyler) [reply server/core/server_test.go]: commited, pushed, squashed

File: server/core/stan2kafka.go (hunk 8 above)
Comment (lmtyler) [reply server/core/stan2kafka.go:23]: commited, pushed, squashed

**Review by derekcollison — APPROVED** [review]
LGTM

**Pull request merged**